### PR TITLE
[Hardware][Intel-Gaudi] Add Intel Gaudi (HPU) inference backend

### PR DESCRIPTION
**Pull request by kzawora-intel**
This PR adds initial support for Intel Gaudi backend to vLLM. 

Requirements
------------

-   OS: Ubuntu 22.04 LTS
-   Python: 3.10
-   Intel Gaudi accelerator
-   Intel Gaudi software version 1.18.0

Supported Features
==================

-   [Offline batched inference](https://github.com/HabanaAI/vllm-fork/blob/habana_main/docs/source/getting_started/quickstart.rst#offline-batched-inference)
-   Online inference via [OpenAI-Compatible  Server](https://github.com/HabanaAI/vllm-fork/blob/habana_main/docs/source/getting_started/quickstart.rst#openai-compatible-server)
-   HPU autodetection
-   Paged KV cache with algorithms enabled for Intel Gaudi accelerators
-   Custom Intel Gaudi implementations of Paged Attention, KV cache ops, prefill attention, Root Mean Square Layer Normalization, Rotary Positional Encoding, FusedMoE
-   Tensor parallelism support for multi-card inference
-   Inference with [HPU Graphs](https://docs.habana.ai/en/latest/PyTorch/Inference_on_PyTorch/Inference_Using_HPU_Graphs.html) for accelerating low-batch latency and throughput

Unsupported Features
====================

-   Beam search
-   LoRA adapters
-   Attention with Linear Biases (ALiBi)
-   Quantization
-   Prefill chunking (mixed-batch inferencing)

Supported Configurations
========================

The following configurations have been validated to be function with Gaudi devices. Configurations that are not listed may or may not work.

- [meta-llama/Llama-2-7b](https://huggingface.co/meta-llama/Llama-2-7b) on single HPU, or with tensor parallelism on 2x and 8x HPU, BF16 datatype with random or greedy sampling
- [meta-llama/Llama-2-7b-chat-hf](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf) on single HPU, or with tensor parallelism on 2x and 8x HPU, BF16 datatype with random or greedy sampling
- [meta-llama/Meta-Llama-3-8B](https://huggingface.co/meta-llama/Meta-Llama-3-8B) on single HPU, or with tensor parallelism on 2x and 8x HPU, BF16 datatype with random or greedy sampling
- [meta-llama/Meta-Llama-3-8B-Instruct](https://huggingface.co/meta-llama/Meta-Llama-3-8B-Instruct) on single HPU, or with tensor parallelism on 2x and 8x HPU, BF16 datatype with random or greedy sampling
- [meta-llama/Meta-Llama-3.1-8B](https://huggingface.co/meta-llama/Meta-Llama-3.1-8B) on single HPU, or with tensor parallelism on 2x and 8x HPU, BF16 datatype with random or greedy sampling
- [meta-llama/Meta-Llama-3.1-8B-Instruct](https://huggingface.co/meta-llama/Meta-Llama-3.1-8B-Instruct) on single HPU, or with tensor parallelism on 2x and 8x HPU, BF16 datatype with random or greedy sampling
- [meta-llama/Llama-2-70b](https://huggingface.co/meta-llama/Llama-2-70b) with tensor parallelism on 8x HPU, BF16 datatype with random or greedy sampling
- [meta-llama/Llama-2-70b-chat-hf](https://huggingface.co/meta-llama/Llama-2-70b-chat-hf) with tensor parallelism on 8x HPU, BF16 datatype with random or greedy sampling
- [meta-llama/Meta-Llama-3-70B](https://huggingface.co/meta-llama/Meta-Llama-3-70B) with tensor parallelism on 8x HPU, BF16 datatype with random or greedy sampling
- [meta-llama/Meta-Llama-3-70B-Instruct](https://huggingface.co/meta-llama/Meta-Llama-3-70B-Instruct) with tensor parallelism on 8x HPU, BF16 datatype with random or greedy sampling
- [meta-llama/Meta-Llama-3.1-70B](https://huggingface.co/meta-llama/Meta-Llama-3.1-70B) with tensor parallelism on 8x HPU, BF16 datatype with random or greedy sampling
- [meta-llama/Meta-Llama-3.1-70B-Instruct](https://huggingface.co/meta-llama/Meta-Llama-3.1-70B-Instruct) with tensor parallelism on 8x HPU, BF16 datatype with random or greedy sampling

Map of changes
==================
- `vllm/executor/hpu_executor.py` - 1xHPU executor inheriting from `ExecutorBase`
- `vllm/executor/ray_hpu_executor.py` - Multi-HPU (2x, 4x, 8x) executor inheriting from `DistributedGPUExecutor`
- `vllm/engine/attention/backends/hpu_attn.py` - Gaudi-specific backend for handling prefill attention and paged attention
- `vllm/engine/attention/ops/hpu_paged_attn.py` - Gaudi backend for handling paged attention
- `vllm/worker/hpu_worker.py` - Gaudi-specific worker for handling distributed inference and executing models, inherited from `WorkerBase`
- `vllm/worker/hpu_model_runner.py` - Gaudi-specific model executor, inherited from `ModelRunnerBase`, with input class `ModelInputForHPU` inherited from `ModelRunnerInputBase`
- `vllm/model_executor/layers/*.py` - Gaudi-specific forward passes for some operators, otherwise falling back to `forward_native`
- `vllm/model_executor/engine/*.py` - Routing logic for dispatching proper Gaudi executors
- Other changes are either minor, non-invasive workarounds for Gaudi specifically, or utilities.
- PR was passed through format.sh with no issues.

### CODE REVIEW

File: vllm/distributed/communication_op.py [hunk 1]
```diff
@@ -3,11 +3,21 @@
...
+    import habana_frameworks.torch as htorch
+
 
 def tensor_model_parallel_all_reduce(input_: torch.Tensor) -> torch.Tensor:
     """All-reduce the input tensor across model parallel group."""
+    if is_hpu():
+        # FIXME(kzawora): this is a workaround for a bug in Habana PT bridge
+        # occurring when PT_HPU_ENABLE_LAZY_COLLECTIVES=true env var is used
+        # (which is required for tensor parallel HPUGraph inference)
+        htorch.core.mark_step()
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/distributed/communication_op.py:20]: Indeed the [TPU CI test](https://buildkite.com/vllm/fastcheck/builds/294#0190baef-921f-4b7d-a7a8-49c2700be409) failed for this reason.

We temporarily stopped the TPU CI test so it will not err if retry, but please remove it from the compiler path.

**Review by WoosukKwon — COMMENTED** [review]
@kzawora-intel Thanks for submitting the PR, and sorry for the delays in the review! Overall, the PR looks clean and is pretty up to date. Really appreciate it.

I did some preliminary reviews, mostly on the changes on the existing files. While I was able to notice that you managed hard to make the PR less intrusive to the current codebase, I feel it can be further modularized. Small `if` statements that might look trivial for someone can be intractable and annoying to others.

Will follow up with more reviews later.

File: vllm/worker/cache_engine.py [hunk 2]
```diff
@@ -6,9 +6,12 @@
 from vllm.attention import get_attn_backend
 from vllm.config import CacheConfig, DeviceConfig, ModelConfig, ParallelConfig
 from vllm.logger import init_logger
-from vllm.utils import (STR_DTYPE_TO_TORCH_DTYPE, get_dtype_size,
+from vllm.utils import (STR_DTYPE_TO_TORCH_DTYPE, get_dtype_size, is_hpu,
                         is_pin_memory_available)
 
+if is_hpu():
+    pass
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/worker/cache_engine.py]: fixed, thanks!

File: vllm/model_executor/layers/logits_processor.py [hunk 3]
```diff
@@ -93,8 +93,11 @@ def _prune_hidden_states(
     hidden_states: torch.Tensor,
     sampling_metadata: SamplingMetadata,
 ) -> torch.Tensor:
-    return hidden_states.index_select(0,
-                                      sampling_metadata.selected_token_indices)
+    if sampling_metadata.selected_token_indices is not None:
+        return hidden_states.index_select(
+            0, sampling_metadata.selected_token_indices)
+    else:
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/model_executor/layers/logits_processor.py:101]: added, thanks!

File: vllm/model_executor/layers/layernorm.py [hunk 4]
```diff
@@ -67,6 +75,43 @@ def forward_cuda(
         )
         return out
 
+    def forward_hpu(
+        self,
+        x: torch.Tensor,
+        residual: Optional[torch.Tensor] = None,
+    ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
+        from vllm._ipex_ops import ipex_ops as ops
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/model_executor/layers/layernorm.py]: good catch, thanks! this was an aciddent, we don't use it - the entire hpu forward pass has been refactored now 😸

File: vllm/model_executor/layers/vocab_parallel_embedding.py [hunk 5]
```diff
@@ -327,8 +328,19 @@ def weight_loader(self, param: Parameter, loaded_weight: torch.Tensor):
 
         # Copy the data.
         loaded_weight = loaded_weight.narrow(output_dim, start_idx, shard_size)
-        param[:loaded_weight.shape[0]].data.copy_(loaded_weight)
-        param[loaded_weight.shape[0]:].data.fill_(0)
+
+        # FIXME(kzawora): Weight copy with slicing bugs out on Gaudi here, so
+        # we're using a workaround. Remove this when fixed in HPU PT bridge.
+        if is_hpu():
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/model_executor/layers/vocab_parallel_embedding.py]: done

File: docs/source/getting_started/gaudi-installation.rst [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+   datatype with random or greedy sampling
+-  `meta-llama/Llama-2-7b-chat-hf <https://huggingface.co/meta-llama/Llama-2-7b-chat-hf>`__
+   on single HPU, or with tensor parallelism on 2x and 8x HPU, BF16
+   datatype with random or greedy sampling
+-  `meta-llama/Llama-2-70b <https://huggingface.co/meta-llama/Llama-2-70b>`__
+   with tensor parallelism on 8x HPU, BF16 datatype with random or
+   greedy sampling
+-  `meta-llama/Llama-2-70b-chat-hf <https://huggingface.co/meta-llama/Llama-2-70b-chat-hf>`__
+   with tensor parallelism 8x HPU, BF16 datatype with random or greedy
+   sampling
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply docs/source/getting_started/gaudi-installation.rst]: I've updated the list with Llama3 - we've confirmed them to work, but 3.1 has issues with RoPE scaling (the same reported in https://github.com/vllm-project/vllm/issues/6689) - will give it a go again after rebasing

File: docs/source/getting_started/gaudi-installation.rst [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+Use the following commands to run a Docker image:
+
+.. code:: console
+
+   $ docker pull vault.habana.ai/gaudi-docker/1.16.0/ubuntu22.04/habanalabs/pytorch-installer-2.2.0:latest
+   $ docker run -it --runtime=habana -e HABANA_VISIBLE_DEVICES=all -e OMPI_MCA_btl_vader_single_copy_mechanism=none --cap-add=sys_nice --net=host --ipc=host vault.habana.ai/gaudi-docker/1.16.0/ubuntu22.04/habanalabs/pytorch-installer-2.2.0:latest
+
+Build and Install vLLM-fork
+---------------------------
+
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply docs/source/getting_started/gaudi-installation.rst:52]: Rephrased the entire section, included installation guide for both upstream and fork - thanks!

File: docs/source/getting_started/gaudi-installation.rst [hunk 8]
```diff
@@ -0,0 +1,144 @@
+vLLM with Intel® Gaudi® 2 AI Accelerators
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply docs/source/getting_started/gaudi-installation.rst]: Not at all - this code should run as well for other Gaudi accelerators 😸 I've removed the number from the title, thanks!

File: docs/source/getting_started/gaudi-installation.rst [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+---------------------------
+
+To build and install vLLM-fork from source, run:
+
+.. code:: console
+
+   $ git clone https://github.com/HabanaAI/vllm-fork.git
+   $ cd vllm-fork
+   # git checkout v0.4.2-Gaudi-1.16.0
+   $ pip install -e .  # This may take 5-10 minutes.
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply docs/source/getting_started/gaudi-installation.rst]: Changed the example to `python setup.py develop`, thanks!. While ` pip install -e .` did not break anything in our environments, it did pull PyTorch with CUDA unnecessarily.

File: docs/source/getting_started/gaudi-installation.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply docs/source/getting_started/gaudi-installation.rst]: After rebasing up to 60d1c6e584203b07df5020529d678e69baad84a3 on https://github.com/HabanaAI/vllm-fork/pull/108, I can confirm that Llama3.1-8B-Instruct is running and the RoPE scaling issue is gone. We'll do some tests internally, and if we observe no major issues, we'll merge the changes and update this PR.

File: vllm/model_executor/models/mixtral.py [hunk 10]
```diff
@@ -432,3 +432,6 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
                     weight_loader = getattr(param, "weight_loader",
                                             default_weight_loader)
                     weight_loader(param, loaded_weight)
+
+            if is_hpu():
+                torch.hpu.synchronize()
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/model_executor/models/mixtral.py:437]: This is a workaround - without synchronization here, loading Mixtral weights resulted in unusually high HPU memory usage, which could end up in OOM before warmup.

File: vllm/distributed/communication_op.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/distributed/communication_op.py:20]: @kzawora-intel Could you please mimic what I did in #6813 to avoid a similar issue in TPU?

File: vllm/hpu/rotary_embed.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+def is_gaudi2():
+    return get_device_type() == htexp.synDeviceType.synDeviceGaudi2
+
+
+def is_gaudi3():
+    return get_device_type() == htexp.synDeviceType.synDeviceGaudi3
+
+
+# TODO: remove this workaround when FusedRoPE properly works on Gaudi
+if not is_gaudi1() and (is_gaudi2() or is_gaudi3()):
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/hpu/rotary_embed.py]: we do now :) removed this check

File: vllm/distributed/communication_op.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/distributed/communication_op.py:20]: https://github.com/HabanaAI/vllm-fork/pull/136 - like so? if that looks good, I'll push it onto this branch

File: vllm/executor/ray_utils.py [hunk 12]
```diff
@@ -85,7 +86,11 @@ def initialize_ray_cluster(
                  ignore_reinit_error=True,
                  num_gpus=parallel_config.world_size)
     else:
-        ray.init(address=ray_address, ignore_reinit_error=True)
+        ray.init(address=ray_address,
+                 ignore_reinit_error=True,
+                 log_to_driver=os.environ.get("VLLM_RAY_DISABLE_LOG_TO_DRIVER",
+                                              "0") == "0")
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/executor/ray_utils.py]: We've used it for development at some point, to filter out logs from ray workers on driver. We don't use it anymore. I've removed it.

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 13]
```diff
@@ -100,6 +104,12 @@ def forward_cuda(
                          num_expert_group=num_expert_group,
                          topk_group=topk_group)
 
+    def forward_hpu(self, x: torch.Tensor, w1: torch.Tensor, w2: torch.Tensor,
+                    router_logits: torch.Tensor, top_k: int, renormalize: bool,
+                    use_grouped_topk: bool, num_expert_group: Optional[int],
+                    topk_group: Optional[int]):
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/model_executor/layers/fused_moe/layer.py:110]: added, thanks!

File: vllm/hpu/rotary_embed.py [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+    try:
+        from habana_frameworks.torch.hpex.kernels import (
+            RotaryPosEmbeddingHelperV1 as FusedRoPE)
+    except ImportError:
+        print("Not using HPU fused kernel for apply_rotary_pos_emb")
+        FusedRoPE = None
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/hpu/rotary_embed.py]: I've refactored the code here - if for whatever reason HPEX import fails, we'll fall back to native implementation. Some earlier Synapse builds used to have issues with RotaryPosEmbeddingHelperV1 on Gaudi1 specifically - this handles that case and if dedicated kernel is available, we'll use that, else we'll fall back.

File: docs/source/getting_started/gaudi-installation.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply docs/source/getting_started/gaudi-installation.rst]: no major issues observed, PR merged, and llama3.1 should be now functional too 🐱

File: vllm/model_executor/model_loader/loader.py [hunk 15]
```diff
@@ -271,7 +271,7 @@ def load_model(self, *, model_config: ModelConfig,
                    scheduler_config: SchedulerConfig,
                    cache_config: CacheConfig) -> nn.Module:
         with set_default_torch_dtype(model_config.dtype):
-            with torch.device(device_config.device):
+            with torch.device(torch.device(device_config.device)):
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/model_executor/model_loader/loader.py]: it is, removed

File: vllm/distributed/communication_op.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/distributed/communication_op.py]: merged

File: vllm/hpu/utils.py [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+###############################################################################
+
+import habana_frameworks.torch as htorch
+
+
+def with_mark_steps(fn):
+
+    def wrapped(*args, **kwargs):
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/hpu/utils.py:16]: fixed, thanks!

File: setup.py [hunk 17]
```diff
@@ -370,6 +405,12 @@ def get_vllm_version() -> str:
         if neuron_version != MAIN_CUDA_VERSION:
             neuron_version_str = neuron_version.replace(".", "")[:3]
             version += f"+neuron{neuron_version_str}"
+    elif _is_hpu():
+        # Get the Intel Gaudi Software Suite version
+        gaudi_sw_version = str(get_gaudi_sw_version())
+        if gaudi_sw_version != MAIN_CUDA_VERSION:
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply setup.py]: fixed, thanks!

File: vllm/model_executor/layers/logits_processor.py [hunk 18]
```diff
@@ -109,8 +109,13 @@ def _prune_hidden_states(
     hidden_states: torch.Tensor,
     sampling_metadata: SamplingMetadata,
 ) -> torch.Tensor:
-    return hidden_states.index_select(0,
-                                      sampling_metadata.selected_token_indices)
+    # NOTE(kzawora): This is needed for Gaudi - in some scenarios (warmup,
```
Comment (WoosukKwon) [inline vllm/model_executor/layers/logits_processor.py:112]:
```suggestion
    # NOTE(kzawora): The if guard is needed for Gaudi - in some scenarios (warmup,
```

File: vllm/model_executor/layers/quantization/inc.py
Comment (WoosukKwon) [inline vllm/model_executor/layers/quantization/inc.py:1]: Can you please move quantization to another PR?

File: vllm/worker/cache_engine.py
Comment (WoosukKwon) [inline vllm/worker/cache_engine.py:1]: Can we not share the cache engine with the GPU backends? Actually, the cache engine only does a few things: allocating the KV cache and call `swap_blocks` or `copy_blocks` when needed (does Gaudi support this?).

File: vllm/model_executor/models/llama.py [hunk 19]
```diff
@@ -324,6 +327,9 @@ def forward(
             hidden_states = intermediate_tensors["hidden_states"]
             residual = intermediate_tensors["residual"]
 
+        if is_hpu:
+            import habana_frameworks.torch as htorch
+            htorch.core.mark_step()
```
Comment (WoosukKwon) [inline vllm/model_executor/models/llama.py:332]: Why do we need this?

File: vllm/model_executor/models/llama.py [hunk 20]
```diff
@@ -545,6 +553,8 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
                 weight_loader = getattr(param, "weight_loader",
                                         default_weight_loader)
                 weight_loader(param, loaded_weight)
+            if current_platform.is_hpu():
+                torch.hpu.synchronize()
```
Comment (WoosukKwon) [inline vllm/model_executor/models/llama.py:557]: Can we remove this hack by hooking into the weight loader? For example, we do a similar thing for TPUs: https://github.com/vllm-project/vllm/blob/7f60520deb05d2e097b408e3310f1d383fbf1de6/vllm/model_executor/model_loader/loader.py#L358

File: vllm/model_executor/layers/linear.py
Comment (WoosukKwon) [inline vllm/model_executor/layers/linear.py:1]: Why do we need this change?

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 21]
```diff
@@ -204,6 +224,9 @@ def __init__(
         self.num_expert_group = num_expert_group
         self.topk_group = topk_group
         self.custom_routing_function = custom_routing_function
+        if current_platform.is_hpu():
+            from vllm_hpu_extension.ops import StaticFusedMOE
+            self.hpu_static_fused_moe = StaticFusedMOE(self.num_experts)
```
Comment (WoosukKwon) [inline vllm/model_executor/layers/fused_moe/layer.py:229]: Could you please elaborate on how this works?

File: vllm/core/block/cpu_gpu_block_allocator.py [hunk 22]
```diff
@@ -52,7 +53,11 @@ def create(
             - The block IDs are assigned contiguously, with GPU block IDs coming
                 before CPU block IDs.
         """
-        block_ids = list(range(num_gpu_blocks + num_cpu_blocks))
+        # For HPU, block id 0 is used only for padding
+        reserved_blocks = 1 if current_platform.is_hpu() else 0
+        block_ids = list(
+            range(reserved_blocks, num_gpu_blocks + num_cpu_blocks))
+        num_gpu_blocks -= reserved_blocks
```
Comment (WoosukKwon) [inline vllm/core/block/cpu_gpu_block_allocator.py:60]: I'm actually not sure whether this change makes sense.

**Review by WoosukKwon — COMMENTED** [review]
Thanks for the PR! Overall, I'm happy that the PR size is smaller now, and only have a few suggestions:

1. Can we move quantization part out to a followup PR?
2. Why do we need changes on the weight loading (e.g., `weights-load-device`)?
3. Why do we need `htorch.mark_step` inside the model?
4. Could you please elaborate on how the MoE works?

File: vllm/core/block/cpu_gpu_block_allocator.py (hunk 22 above)
Comment (WoosukKwon) [reply vllm/core/block/cpu_gpu_block_allocator.py:60]: Let me find someone to ask if this is safe.

File: Dockerfile.hpu [hunk 23]
```diff
@@ -0,0 +1,18 @@
+FROM vault.habana.ai/gaudi-docker/1.17.0/ubuntu22.04/habanalabs/pytorch-installer-2.3.1:latest
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply Dockerfile.hpu:1]: https://github.com/vllm-project/vllm/issues/8932

File: vllm/model_executor/layers/linear.py
Comment (kzawora-intel) [reply vllm/model_executor/layers/linear.py:1]: this was needed for quantization with INC, I've removed this now.

File: vllm/model_executor/models/llama.py (hunk 20 above)
Comment (kzawora-intel) [reply vllm/model_executor/models/llama.py]: I've removed this workaround completely for now - it was needed only for Llama-405B

File: vllm/model_executor/models/llama.py (hunk 19 above)
Comment (kzawora-intel) [reply vllm/model_executor/models/llama.py]: This is an optimization to cut down on graph compile time. It does not change functionality or throughput on compiled graph. I've removed it now.

File: vllm/worker/habana_worker.py
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/worker/habana_worker.py:1]: renamed, thanks!

File: vllm/worker/habana_model_runner.py
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/worker/habana_model_runner.py:1]: renamed, thanks!

File: vllm/model_executor/layers/quantization/inc.py
Comment (kzawora-intel) [reply vllm/model_executor/layers/quantization/inc.py:1]: I removed quantization from this PR

File: vllm/model_executor/layers/rotary_embedding.py [hunk 24]
```diff
@@ -917,8 +921,16 @@ def get_rope(
         return _ROPE_DICT[key]
 
     if rope_scaling is None:
-        rotary_emb = RotaryEmbedding(head_size, rotary_dim, max_position, base,
-                                     is_neox_style, dtype)
+        if current_platform.is_hpu():
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply vllm/model_executor/layers/rotary_embedding.py:921]: done

File: vllm/worker/model_runner_base.py [hunk 25]
```diff
@@ -47,9 +47,14 @@ def _init_attn_metadata_from_tensor_dict(
...
     for field in dataclasses.fields(attn_backend.get_metadata_cls()):
         val = tensor_dict.pop(field.name, None)
-        if val is not None:
+        # NOTE(kzawora): None is a valid value if type is optional. If
+        # we don't check against it, we will crash by not assigning
+        # Optional types without default value, even if they are
+        # broadcasted properly.
+        is_field_optional = get_origin(field.type) is Union and \
+            type(None) in get_args(field.type)
+        if val is not None or (val is None and is_field_optional):
```
Comment (njhill) [inline vllm/worker/model_runner_base.py:56]: Could we not just add detault `None` values to the relevant fields in `HPUAttentionMetadata`? There may be some cost to doing these checks every time.

Alternatively could use some other sentinel value for the `pop` default, assuming `tensor_dict` will contain actual `None` valued entries in these cases?

Or if neither of those are possible for some reason, let's only compute `is_field_optional` in the case that `val` is `None` ....

**Review by njhill — COMMENTED** [review]
Thanks for all of the work on this @kzawora-intel!

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 21 above)
Comment (jbyczkow) [reply vllm/model_executor/layers/fused_moe/layer.py:229]: This is a MoE implementation that contains only static shapes at the cost of tensors padding. When tokens are allocated to experts, tensors shapes are not determined by the specific allocation, instead they are aligned to token count. This was proposed to avoid the need to handle dynamic shapes for hpu implementation.

File: vllm/worker/model_runner_base.py (hunk 25 above)
Comment (kzawora-intel) [reply vllm/worker/model_runner_base.py:56]: Great idea with the sentinel! Changed now.

File: Dockerfile.hpu [hunk 26]
```diff
@@ -0,0 +1,18 @@
...
+RUN pip install -v -r requirements-hpu.txt
+
+ENV no_proxy=localhost,127.0.0.1
+ENV PT_HPU_ENABLE_LAZY_COLLECTIVES=true
+
+RUN VLLM_TARGET_DEVICE=hpu python3 setup.py install
+
+WORKDIR /workspace/
+
+RUN ln -s /workspace/vllm/tests && ln -s /workspace/vllm/examples && ln -s /workspace/vllm/benchmarks
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply Dockerfile.hpu]: this was actually not needed, removed now

File: Dockerfile.hpu (hunk 23 above)
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply Dockerfile.hpu]: Image with 2.4 is now here, 2.5 will be also supported shortly.

File: vllm/model_executor/model_loader/loader.py [hunk 27]
```diff
@@ -98,8 +98,8 @@ def _get_quantization_config(
     """Get the quantization config."""
     if model_config.quantization is not None:
         quant_config = get_quant_config(model_config, load_config)
-        capability_tuple = current_platform.get_device_capability()
-
+        capability_tuple = current_platform.get_device_capability() \
+                            if current_platform.is_cuda_alike() else None
```
Comment (WoosukKwon) [inline vllm/model_executor/model_loader/loader.py:102]: Please revert this as we decided to add quantization in the next PR.

File: vllm/model_executor/layers/quantization/utils/w8a8_utils.py [hunk 28]
```diff
@@ -121,7 +121,7 @@ def apply_fp8_linear(
         qinput, x_scale = ops.scaled_fp8_quant(
             input,
             input_scale,
-            num_token_padding=17,
+            batch_dim_padding=17,
```
Comment (WoosukKwon) [inline vllm/model_executor/layers/quantization/utils/w8a8_utils.py:124]: Please revert this as we decided to add quantization in the next PR.

File: vllm/engine/multiprocessing/engine.py [hunk 29]
```diff
@@ -209,7 +209,7 @@ def run_engine_loop(self):
             self._alive()
             if not self.engine.has_unfinished_requests():
                 # Poll until there is work to do.
-                while self.input_socket.poll(timeout=POLLING_TIMEOUT_MS) == 0:
+                while self.input_socket.poll(timeout=VLLM_RPC_TIMEOUT) == 0:
```
Comment (WoosukKwon) [inline vllm/engine/multiprocessing/engine.py:212]: Do we need this change?

File: vllm/engine/async_llm_engine.py [hunk 30]
```diff
@@ -1294,15 +1303,17 @@ def remove_logger(self, logger_name: str) -> None:
     async def start_profile(self) -> None:
         # using type instead of isinstance to check to avoid capturing
         # inherited classes
-        if type(self.engine.model_executor) == GPUExecutorAsync:  # noqa: E721
+        if type(self.engine.model_executor) == GPUExecutorAsync or \
+            type(self.engine.model_executor) == HPUExecutorAsync:  # noqa: E721
```
Comment (WoosukKwon) [inline vllm/engine/async_llm_engine.py:1307]: Please revert this change.

File: vllm/engine/llm_engine.py [hunk 31]
```diff
@@ -1767,15 +1776,17 @@ def check_health(self) -> None:
     def start_profile(self) -> None:
         # using type instead of isinstance to check to avoid capturing
         # inherited classes (MultiprocessingGPUExecutor)
-        if type(self.model_executor) == GPUExecutor:  # noqa: E721
+        if type(self.model_executor) == GPUExecutor or \
```
Comment (WoosukKwon) [inline vllm/engine/llm_engine.py:1779]: ditto

File: vllm/model_executor/layers/rotary_embedding.py [hunk 32]
```diff
@@ -917,8 +918,17 @@ def get_rope(
...
+            from vllm_hpu_extension.rotary_embed import HpuRotaryEmbedding
+            rotary_emb = HpuRotaryEmbedding(head_size,
+                                            rotary_dim,
+                                            max_position,
+                                            base,
+                                            is_neox_style,
+                                            RoPEFallback=RotaryEmbedding)
```
Comment (WoosukKwon) [inline vllm/model_executor/layers/rotary_embedding.py:928]: QQ: Can we implement this in `forward_hpu` of `RotaryEmbedding`?

File: vllm/engine/llm_engine.py (hunk 31 above)
Comment (kzawora-intel) [reply vllm/engine/llm_engine.py]: reverted

File: vllm/engine/multiprocessing/engine.py (hunk 29 above)
Comment (kzawora-intel) [reply vllm/engine/multiprocessing/engine.py]: removed

File: vllm/model_executor/layers/quantization/utils/w8a8_utils.py (hunk 28 above)
Comment (kzawora-intel) [reply vllm/model_executor/layers/quantization/utils/w8a8_utils.py]: removed

File: vllm/model_executor/model_loader/loader.py (hunk 27 above)
Comment (kzawora-intel) [reply vllm/model_executor/model_loader/loader.py]: removed

File: vllm/engine/llm_engine.py (hunk 31 above)
Comment (kzawora-intel) [reply vllm/engine/llm_engine.py]: removed

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 21 above)
Comment (kzawora-intel) [reply vllm/model_executor/layers/fused_moe/layer.py]: We agreed to drop StaticFusedMOE from this PR's scope.

File: vllm/worker/cache_engine.py
Comment (kzawora-intel) [reply vllm/worker/cache_engine.py:1]: Removed HPU code from cache engine - I've created a HPUCacheEngine inheriting from regular CacheEngine and just overriding the cache allocator. It is defined in the worker.

File: vllm/model_executor/layers/rotary_embedding.py (hunk 32 above)
Comment (kzawora-intel) [reply vllm/model_executor/layers/rotary_embedding.py]: sure, just did 😺 The changes in RoPE are now limited to forward_hpu

File: requirements-hpu.txt [hunk 33]
```diff
@@ -0,0 +1,8 @@
+# Common dependencies
+-r requirements-common.txt
+
+# Dependencies for HPU code
+ray == 2.32.0
```
[the remark this answers is not in the document]
Comment (kzawora-intel) [reply requirements-hpu.txt]: removed, this was not needed

**Review by WoosukKwon — APPROVED** [review]
LGTM! Thanks for addressing all the comments! Let me do the final checks on whether the PR affects the perf/functionality of other hardware backends, before the PR is merged.

File: docs/source/index.rst [hunk 34]
```diff
@@ -43,7 +43,7 @@ vLLM is flexible and easy to use with:
 * Tensor parallelism and pipeline parallelism support for distributed inference
 * Streaming outputs
 * OpenAI-compatible API server
-* Support NVIDIA GPUs, AMD CPUs and GPUs, Intel CPUs and GPUs, PowerPC CPUs, TPU, and AWS Trainium and Inferentia Accelerators.
+* Support NVIDIA GPUs, AMD CPUs and GPUs, Intel CPUs and Gaudi® accelerators, GPUs, PowerPC CPUs, TPU, and AWS Trainium and Inferentia Accelerators.
```
Comment (zhouyuan) [inline docs/source/index.rst:46]:
```suggestion
* Support NVIDIA GPUs, AMD CPUs and GPUs, Intel CPUs, Gaudi® accelerators and GPUs, PowerPC CPUs, TPU, and AWS Trainium and Inferentia Accelerators.
```

**Review by zhouyuan — COMMENTED** [review]
Just one nit on the supported intel devices

**Comment by kzawora-intel** [comment]
@WoosukKwon PR #9938 introduced some API mismatches in workers/model_runners, I've updated this PR to reflect that change.

**Comment by WoosukKwon** [comment]
@kzawora-intel Just merged the PR. Really appreciate all of you & your team's efforts on the PR. Congrats!! 🎉

**Comment by warlock135** [comment]
@kzawora-intel The file vllm/model_executor/sampling_metadata.py should also be merged (or applying vllm-fork PR #246).
Currently, vLLM raises the following error when using the HPU inference backend: "RuntimeError: Device type HPU is not supported for torch.Generator() API"
